File: app/src/main/java/com/liang/tind/leetcode/backtracking/LeetCode78.java
package com.liang.tind.leetcode.backtracking;

import java.util.ArrayList;
import java.util.List;

/**
 * created by sherlock
 * <p>
 * date 2019/12/22
 * <p>
 * 给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
 * <p>
 * 说明：解集不能包含重复的子集。
 * <p>
 * 示例:
 * <p>
 * 输入: nums = [1,2,3]
 * 输出:
 * [
 * [3],
 *   [1],
 *   [2],
 *   [1,2,3],
 *   [1,3],
 *   [2,3],
 *   [1,2],
 *   []
 * ]
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/subsets
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class LeetCode78 {

    /**
     * 结束条件是 cur== nums.length;
     * 路径：tacks
     * <p>
     * dfs()
     */
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(0, nums, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int cur, int[] nums, List<Integer> tracks, List<List<Integer>> res) {
        /**
         *
         *[[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]]
         */
        res.add(new ArrayList<Integer>(tracks));
        for (int i = cur; i < nums.length; i++) {
            tracks.add(nums[i]);
            dfs(i+1, nums, tracks, res);
            tracks.remove(tracks.size() - 1);
        }
    }

}

File: app/src/main/java/com/liang/tind/leetcode/backtrace/LeetCode78.kt
package com.liang.tind.leetcode.backtrace

import java.util.*
import java.util.stream.Collectors

/**
 *@author: lonnie.liang
 *date: 2022/05/20 23:32
 *
 */
class LeetCode78 {
    private val results: MutableList<List<Int>> = LinkedList()

    fun subsets(nums: IntArray): List<List<Int>> {


        // solution1(nums)
        solution2(nums)
        return results
    }

    private fun solution1(nums: IntArray) {
        results.clear()
        // 空集
        results.add(LinkedList())
        for (num in nums) {
            // 遍历nums 把之前结果中的每一个子集 都 copy出来 然后add num， 然后把copy 重新添加回去results中
            // 初始： results = [[]]
            // 第一轮：results[[]], subList = [[1]] map结束以后 -> results=[[], [1]]
            // 第二轮：results=[[], [1]], subList = [[2], [1,2]] map结束以后 -> results=[[], [1], [2], [1,2]]
            val subList = results.map { subset: List<Int> ->
                val newSubset: MutableList<Int> = LinkedList(subset)
                newSubset.add(num)
                newSubset
            }

            results.addAll(subList)
        }
    }

    private fun solution2(nums: IntArray) {
        results.clear()
        backStrace(0, mutableListOf(), nums)
    }
    private fun backStrace(cur: Int, track:MutableList<Int>, nums: IntArray) {
        //         [ ]
        //     [1]    [2] [3]
        //   [2] [3] [3]
        // [3]
        // 决策树入行，遍历可以当前选择列表，进入节点先把当前走过的路径添加进去

        results.add(ArrayList(track))
        for (i in cur until nums.size) {
            track.add(nums[i])
            backStrace(i+1, track, nums)
            track.removeAt(track.lastIndex)
        }
    }
}